asr1k_neutron_l3/models/netconf_yang/l3_interface.py

```
@@ -500,3 +523,179 @@ def to_dict(self, context):
+        if self.keepalive_period:
+            keepalive[L3Constants.PERIOD] = self.keepalive_period
+        if self.keepalive_retries:
+            keepalive[L3Constants.RETRIES] = self.keepalive_retries
+        if keepalive:
+            iface[L3Constants.KEEPALIVE_CONFIG] = keepalive
+
+        ip = {}
+        if self.ipv4_address:
+            ip[L3Constants.ADDRESS] = {
+                L3Constants.PRIMARY: {
+                    L3Constants.ADDRESS: self.ipv4_address,
+                    L3Constants.MASK: self.ipv4_netmask,
+                }
+            }
+        if self.tcp_mss:
+            ip[L3Constants.TCP] = {L3Constants.ADJUST_MSS: self.tcp_mss}
+        if self.mtu:
+            ip[L3Constants.MTU] = self.mtu
+        if ip:
+            iface[L3Constants.IP] = ip
+
+        ipv6 = {}
+        if self.ipv6_prefix:
+            ipv6[L3Constants.ADDRESS] = {L3Constants.PREFIX_LIST: {L3Constants.PREFIX: self.ipv6_prefix}}
+        if self.ipv6_tcp_mss:
+            ipv6[L3Constants.TCP] = {L3Constants.ADJUST_MSS: self.ipv6_tcp_mss}
+        if self.ipv6_mtu:
+            ipv6[L3Constants.MTU] = self.ipv6_mtu
+        if ipv6:
+            iface[L3Constants.IPV6] = ipv6
+
+        if self.vrf:
+            iface[L3Constants.VRF] = {L3Constants.FORWARDING: self.vrf}
+
+        if self.shutdown:
+            iface[L3Constants.SHUTDOWN] = ''
+        else:
+            iface[L3Constants.SHUTDOWN] = {xml_utils.OPERATION: NC_OPERATION.REMOVE}
+
+        tun = {}
+        if self.tunnel_src:
+            tun[L3Constants.SOURCE] = self.tunnel_src
+        if self.tunnel_dest_ipv4 or self.tunnel_dest_ipv6:
+            dest_config = {}
+            if self.tunnel_dest_ipv4:
+                dest_config[L3Constants.IPV4] = self.tunnel_dest_ipv4
+            if self.tunnel_dest_ipv6:
+                dest_config[L3Constants.IPV6] = self.tunnel_dest_ipv6
+            tun[L3Constants.DESTINATION_CONFIG] = dest_config
+        if self.tunnel_mode_ipsec_ipv4:
```

Only one of these tunnel modes can be active at a time, I get that (the key is overwritten) but this also logs no error / raises an exception when for some reason the object is broken and the last one active wins. But I also get that there is no validation done at this point because that should have happened earlier already. But there is a chance that this hides an error? (nitpick)

---

The class should be used in a way that only one mode is active, but we always have two data sources. First data source is models.neutron.l3.router.Router / some related object, which is filled with OpenStack info. Due to the way the API works we always should only have one of the two values in here and it should be fine. The other source is the device. If the device shows me in its YANG output that both tunnel modes are there, I want to see this in the model, so that once we do a `router-diff` on it we'll see that there's something different on the device. Both should not be possible from YANG side, but well.... We've both seen our fair share of software :)

---

>  We've both seen our fair share of software :)

does that mean you are also in favor of logging an error or throwing an exception here when more than one is set?